sch_yang_condition.c

```diff
@@ -0,0 +1,923 @@
+/**
+ * @file sch_yang_condition.c
+ * YANG extression processing
+ *
+ * Copyright 2024, Allied Telesis Labs New Zealand, Ltd
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this library. If not, see <http://www.gnu.org/licenses/>
+ */
+
+#include "sch_xpath.h"
+#include <apteryx.h>
+#include <apteryx-xml.h>
+#include <libxml/tree.h>
+#define APTERYX_XML_LIBXML2
+
+
+typedef struct _yang_result
+{
+    bool result;
+    char *value;
+} yang_result;
+
+typedef struct _yang_exists
+{
+    bool match;
+    char *path;
+} yang_exists;
+
+typedef enum
+{
+    PROC_F_IF_FEATURE = (1 << 0),  /* Parent is a if-feature */
+} proc_flags;
+
+
+static void sch_yang_process_xnode (sch_instance *instance, GNode *root, char *path,
+                                    char *step_path, xpath_node *xnode,
+                                    yang_result *presult, int depth, int *flags);
+
+
+static void
+sch_axis_child (char *path, char *child_name, yang_result *presult, int flags)
+{
+    if (child_name)
+    {
+        if ((flags & PROC_F_IF_FEATURE))
+            presult->value = g_strdup (child_name);
+        else
+            presult->value = g_strdup_printf ("%s/%s", path, child_name);
+
+        presult->result = true;
+    }
+    else
+    {
+        presult->result = false;
+    }
+}
+
+static void
+sch_axis_parent (char *path, char *step_path, yang_result *presult, bool self, int flags)
+{
+    char **split;
+    int count;
+    bool same = (!step_path || g_strcmp0 (path, step_path) == 0);
+
+    presult->result = false;
+    if (self && step_path && !same)
+    {
+        presult->result = true;
+        presult->value = g_strdup (step_path);
+        return;
+    }
+
+    if (step_path)
+        split = g_strsplit (step_path, "/", -1);
+    else
+        split = g_strsplit (path, "/", -1);
+
+    if (split)
```

Comment:
I know performance is not particularly top of mind working through this stuff, but wouldn't a simple pass through the string, looking for the last '/' and truncating and copying the string at that point be a lot simpler and quicker? Put this in a routine if such a thing doesn't exist in a standard library, it might be useful in other places.
char *str_truncate_at_last (char * str, char trunc_char);
Also, need to consider what happens if the string starts with '/', or is just '/'.
Also, some comment on what the path/step-path logic is trying to achieve would be helpful.

Reply:
Refactored the routine sch_axis_parent to not use split and join. It now just removes the last path directive.